Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,413 @@
+        ///  |    |    |    } frustumHeight = cameraWindowHeight / 2
+        ///  |---------+  -/
+        ///  |    |    |
+        ///  |---------|
+        /// </summary>
+        private float CalculateFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                Quaternion inverseRotation = Quaternion.Inverse(m_BoundingShape2D.transform.rotation);
+                Vector3 planePosition = inverseRotation * m_BoundingShape2D.transform.position;
+                Vector3 cameraPosition = inverseRotation * vcam.transform.position;
+                float distance = Mathf.Abs(planePosition.z - cameraPosition.z);
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+            return frustumHeight;
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine)
+        {
+            if (ShrinkablePolygon.IsInside(m_currentPathCache, positionToConfine))
+            {
+                return Vector2.zero;
+            }
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < m_currentPathCache.Count; ++i)
+            {
+                int numPoints = m_currentPathCache[i].Count;
+                if (numPoints > 0)
+                {
+                    Vector2 v0 = m_currentPathCache[i][numPoints - 1];
+                    for (int j = 0; j < numPoints; ++j)
+                    {
+                        Vector2 v = m_currentPathCache[i][j];
+                        Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                        float distance = Vector2.SqrMagnitude(positionToConfine - c);
+                        if (distance < minDistance)
+                        {
+                            minDistance = distance;
+                            closest = c;
+                        }
+                        v0 = v;
+                    }
+                }
+            }
+            return closest - positionToConfine;
+        }
+        
+        private class VcamExtraState
+        {
+            public Vector3 m_previousDisplacement;
+            public float confinerDisplacement;
+            public bool applyAfterAim;
+        };
+
+        private float m_sensorRatioCache;
+        private Vector3 m_boundingShapePositionCache;
+        private Vector3 m_boundingShapeScaleCache;
+        private Quaternion m_boundingShapeRotationCache;
```

Comment:
Why don't you just cache a Matrix4x4?

Reply:
Even better (but perhaps a bigger change):  the oven and all the caches could be in bounding shape local space.  That way the confining shape can move around without affecting the cache.  To confine, transform the input point to local space, confine it, then transform back.  Not sure what to do about roll, however...

Reply:
What do you think about doing the calculations in confiner-local space?

Reply:
I think that this would slow down the runtime speed, because we would need to apply Transformation matrix to the points every frame. Right?

What would we gain?